basedtyping/__init__.py

```diff
@@ -556,3 +558,44 @@ def f[T](t: TypeForm[T]) -> T: ...
                             
                              reveal_type(f(int | str))  # int | str
                          """)
+
+
+@dataclass
+class _BaseGenericFunction(Generic[P, T]):
+    fn: Callable[P, T]
+
+
+@dataclass
+class _GenericFunction(_BaseGenericFunction[P, T]):
+    # TODO: make this an TypeVarTuple when mypy supports it
+    #  https://github.com/python/mypy/issues/16696
+    parameters: tuple[object, ...] | None = None
+
+    def __getitem__(self, items: object) -> _ConcreteFunction[P, T]:
+        items = items if isinstance(items, tuple) else (items,)
+        return _ConcreteFunction(self.fn, items)
+
+
+@dataclass
+class _ConcreteFunction(_BaseGenericFunction[P, T]):
+    parameters: tuple[object, ...] | None = None
+
+    def __call__(self, *args: P.args, **kwargs: P.kwargs) -> T:
+        return self.fn(*args, **kwargs)
+
+
+class _GenericFunctionDecorator:
+    args: tuple[object, ...] | None = None
+
+    def __call__(self, fn: Callable[P, T]) -> _GenericFunction[P, T]:
+        return _GenericFunction(fn, self.args)
+
+    def __getitem__(self, items: object) -> Self:
+        items = items if isinstance(items, tuple) else (items,)
+        if self.args:
+            raise RuntimeError("erm, what the spruce?")
```

Comment:
this should probably say what the issue is, or at least include the args in the error message